fix(core): add hydration protected elements

arturovt opened the pull request:
This commit updates the implementation of the DOM renderer and its `setAttribute` method. With these new changes, it checks whether the attribute value is the same before setting it. This resolves issues related to `src` attributes that may be set again during client-side rendering (after the content has been initially rendered on the server). Setting `src` multiple times, even with the same values, can force the browser to load the resource again.

arturovt committed:
fix(core): add hydration protected elements

With this commit, we are now able to handle hydration-protected elements and their attributes.

This enhancement is aimed at improving UX and reducing the number of HTTP requests made by browsers
to external resources during the hydration process. Take, for example, the `iframe` element with a
static `src` attribute, which is rendered on the server. When we begin hydrating that element on the
client, it attempts to set up the static attributes again. This includes setting the `src` attribute
again with the same value it already has. Browsers interpret each change to the `src` attribute as a
new request to load the specified resource. Consequently, the video may flicker as the browser attempts
to reload it.

We now maintain a map of elements that should be protected from hydration. Each element (acting as a key)
maps to a list of attributes, for instance, `iframe -> ['src']`. When we look up an existing element, we
check its tag name and if there's a list of attributes associated with it. If we encounter an `iframe`
(or other protected element), we iterate over its protected attributes and check whether that element
already has those attributes set. If it does, we save this information in the hydration info map to use
it when calling `setAttributes`.

arturovt commented:
Here's the list of valid and invalid cases: elements marked in red reload resources when hydrated. We don't consider cases when the cache is disabled through devtools or the remote resource returns a 302 status code.

- ✅

```html
<img src="/assets/img_girl.jpg" />
```

- ✅

```html
<picture>
  <source media="(min-width: 650px)" srcset="/assets/img_food.jpg" />
  <source media="(min-width: 465px)" srcset="/assets/img_car.jpg" />
  <img src="/assets/img_girl.jpg" style="width: auto" />
</picture>
```

- ✅

```html
<svg width="90" height="90">
  <image xlink:href="/assets/angular.svg" width="90" height="90" />
</svg>
```

- ✅

```html
<video width="400" controls autoplay>
  <source src="/assets/mov_bbb.mp4" type="video/mp4" />
  <source src="/assets/mov_bbb.ogg" type="video/ogg" />
</video>
```

- ✅

```html
<audio controls src="/assets/t-rex-roar.mp3"></audio>
```

- ✅

```html
<video controls src="/assets/friday.mp4">
  <track default kind="captions" srclang="en" src="/assets/friday.vtt" />
</video>
```

---

- ❌

```html
<object
  type="application/pdf"
  data="/assets/In-CC0.pdf"
  width="250"
  height="200"
></object>
```

- ❌

```html
<embed type="video/webm" src="/assets/flower.mp4" width="250" height="200" />
```

- ❌

```html
<iframe src="https://www.youtube-nocookie.com/embed/7HVMn1TpXKQ"></iframe>
```

AndrewKushnir commented:
@arturovt thanks for additional information, this is very helpful. We can extend the `isHydrationProtectedAttr` function described in my previous comment to support `<object>` and `<embed>` cases as well. Please let us know if you might be interested in updating the PR as described in https://github.com/angular/angular/pull/52478#issuecomment-2015938520 and also let us know if you have any questions about the mentioned comment.

rhutchison commented:
Thanks @arturovt and @AndrewKushnir for your efforts on this.

AndrewKushnir commented:
@arturovt thanks for implementing this 👍 I've added this PR to the v18 milestone.

If you get a chance, could you please try to find corresponding sections of HTML spec that talk about this specific behavior of the `<iframe>`, `<object>` and `<embed>` elements, so that we can include it as comments into the code for future reference?

arturovt commented:
@AndrewKushnir I went through the HTML standard and I may guess it doesn't explicitly mandate this behavior or provide detailed instructions on how user agents should handle this aspect of network requests. Because if we talk about `img.src` - setting the `src` attribute to the same value again on the `img` element doesn't trigger resource reload and it's a common behavior implemented by all user agents. Considering the `iframe.src` case, I **may guess** that _the content of an iframe can be dynamic and it may contain interactive elements or scripts that require reloading when the source URL changes_ (took this statement from Google search).

Out of curiosity, wanna check what engines do internally...

AndrewKushnir commented on packages/core/src/render3/util/attrs_utils_with_hydration_support.ts:
```diff
@@ -0,0 +1,108 @@
+    // If the index of the currently hydrating element is found in the `protectedAttributes` map, it
+    // indicates that this element has hydration-protected attributes (for example, if the `element`
+    // is an `iframe` with its `src` attribute set).
+    const protectedAttributeValue = hydrationInfo.protectedAttributes.get(nodeIndex)!;
+    if (attributeValue === protectedAttributeValue) {
+      // We do not want to reset the value since it may negatively affect performance and UX. For
+      // example, in the `<iframe src="...">` case, this would cause a reload within the iframe.
+      return;
```
@arturovt I was thinking if we can protect an attribute value for the first "write" operation only and remove a corresponding key from the map afterwards. This should remove the need to pass the `isFirstPass` flag as an argument. What do you think?
```suggestion
    const { protectedAttributes } = hydrationInfo;
    const protectedAttributeValue = protectedAttributes.get(nodeIndex)!;
    if (attributeValue === protectedAttributeValue) {
      // An initial value is protected for the first `setAttribute` call, further protection would
      // happen as a part of the change detection mechanisms. Remove the node index entry
      // from the map to indicate that no special logic are required for this element and a given
      // attribute.
      protectedAttributes.delete(nodeIndex);
      
      // We do not want to reset the value since it may negatively affect performance and UX. For
      // example, in the `<iframe src="...">` case, this would cause a reload within the iframe.
      return;
```

arturovt replied:
Makes sense, I've updated this code path.

AndrewKushnir replied:
@arturovt I still see the `isFirstPass` being used in the code of this PR. Please let me know once the PR is updated, I can take another look.

AndrewKushnir commented on packages/platform-server/test/hydration_spec.ts:
```diff
@@ -7246,5 +7247,80 @@ describe('platform-server hydration integration', () => {
+            const appRef = await renderAndHydrate(doc, html, SimpleComponent);
+            appRef.tick();
+
+            // Question: How can I create a `setAttribute` spy in these tests? The code above
```
We have a similar logic in the NgOptimizedImage tests:
https://github.com/angular/angular/blob/main/packages/common/test/directives/ng_optimized_image_spec.ts#L61-L66

If that doesn't work, you may need to add a spy for the `setAttribute` method on the `Renderer2` class or provide an overridden `Renderer2` instance during the test setup. There are some example of this in the repo, please let me know if any additional info is required.

AndrewKushnir commented on packages/core/src/render3/util/attrs_utils_with_hydration_support.ts:
```diff
@@ -0,0 +1,108 @@
+import {RElement} from '../interfaces/renderer_dom';
+
+/**
+ * This represents a mapping of elements to an attribute that is hydration protected.
+ * Values are implemented as a list in case they need to be expanded in the future to accommodate
+ * more cases.
```
```suggestion
 * This represents a mapping of element tag name to an attribute name that should not be reset
 * during hydration in case the value is the same.
 * Values are implemented as a list in case they need to be expanded in the future to accommodate
 * more cases.
```

arturovt replied:
Updated.

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1692,7 +1707,16 @@ export function setElementAttribute(
       sanitizer == null ? renderStringify(value) : sanitizer(value, tagName || '', name);
 
-    renderer.setAttribute(element, name, strValue as string, namespace);
+    setAttributeWithHydrationSupport(
```
I'd propose renaming this method to something like this:
```suggestion
    _setAttributeImpl(
```
This should make it more clear that the implementation is tree-shakable and hydration code paths are not invoked for client-only apps.

arturovt replied:
Updated.

AndrewKushnir reviewed (commented):
@arturovt thanks for the PR, the code looks great! I've just left a few comments with some proposals.

Could you please also mark this PR as "Ready for review" via GitHub UI?

arturovt commented:
> Could you please also mark this PR as "Ready for review" via GitHub UI?

Done.

Symbol tests are failing, I will update symbols later. Gotta play with unit tests first to spy on the setAttribute.

AndrewKushnir commented on packages/core/src/render3/instructions/element.ts:
```diff
@@ -294,6 +294,17 @@ function locateOrCreateElementNodeImpl(
+  // `<iframe>`) and it has a set attribute (e.g., `src`).
+  if (hydrationProtectedAttribute && native.hasAttribute(hydrationProtectedAttribute)) {
+    const attributeValue = native.getAttribute(hydrationProtectedAttribute);
+    if (attributeValue) {
```
We may have an empty string here, which would return false. Should we just store any value that we've received from the `getAttribute` call?

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1692,7 +1704,7 @@ export function setElementAttribute(
       sanitizer == null ? renderStringify(value) : sanitizer(value, tagName || '', name);
 
-    renderer.setAttribute(element, name, strValue as string, namespace);
+    _setAttributeImpl(hydrationInfo, renderer, nodeIndex, element, name, strValue, namespace);
```
nit: I'd propose moving hydration info to the very end of the argument list for consistency with other functions that we've updated in this PR.

AndrewKushnir commented on packages/core/src/render3/util/attrs_utils_with_hydration_support.ts:
```diff
@@ -0,0 +1,107 @@
+  attributeName: string,
+  attributeValue: string,
+  namespace?: string | null,
+) {
+  if (
```
For code size optimization, we can extract `hydrationInfo?.protectedAttributes` into a separate const and use this const in the code.
```suggestion
) {
  const protectedAttributes = hydrationInfo?.protectedAttributes;
  if (
```

arturovt commented:
@AndrewKushnir I have address your comments and also update unit tests with spying on the `setAttribute` of the renderer.

AndrewKushnir commented on packages/platform-server/test/hydration_spec.ts:
```diff
@@ -7246,5 +7250,137 @@ describe('platform-server hydration integration', () => {
+              undefined,
+            );
+
+            // Assert that the spy for the hydration-protected element renderer has not been called
+            // at all.
+            expect(setAttributeSpies[1]).not.toHaveBeenCalled();
```
Could you please also add some logic here to change the `value` and make sure that the `setAttribite` was invoked as needed (i.e. that the protection logic doesn't prevent subsequent value updates)?

AndrewKushnir reviewed (commented):
@arturovt thanks for addressing the feedback! I've left a comment about some extra testing cases.

arturovt commented:
@AndrewKushnir your last comment helped me realize I was moving in the wrong direction. We only covered the attribute case (`<iframe [attr.src]="..."`), but `<iframe [src]=value` actually calls `elementPropertyInternal` and then the renderer's `setProperty` method. Do we also need to add the same for properties?

AndrewKushnir commented:
@arturovt thanks for updating the PR and covering additional scenarios!

> Wondering whether we have to update the setProperty implementation too?

Yes, you are right, we may see a similar issue in element property code. I've looked at the existing code and we'd need to add some amount of complexity there to handle that. Let me think about possible options a bit more and I'll reply back with some ideas.

AndrewKushnir commented on packages/core/src/render3/util/attrs_utils_with_hydration_support.ts:
```diff
@@ -0,0 +1,106 @@
+  attributeValue: string,
+  namespace: string | null | undefined,
+  hydrationInfo: DehydratedView | null,
+) {
+  const protectedAttributes = hydrationInfo?.protectedAttributes;
```
Let's wrap extra logic in this function into a check that ensures that we only do that for the first update pass, so subsequent update passes would not pay extra execution cost.
```
    if (lView[FLAGS] & LViewFlags.FirstLViewPass) {
      const protectedAttributes = hydrationInfo?.protectedAttributes;
      // ...
    }
    renderer.setAttribute(element, attributeName, attributeValue, namespace);
```

AndrewKushnir reviewed (commented):
@arturovt I've left a comment with a proposal to check whether we are doing a first LView pass (during this pass, we want to invoke extra logic), which should avoid extra overhead.

Also as you've mentioned, we'd need to handle setting property cases as well. When you get a chance, could you please update this PR to include this logic too?

Thank you.

arturovt commented:
@AndrewKushnir I've updated the code. I created the `_setPropertyImpl` function, which behaves similarly to `_setAttributeImpl`. Since the code for checking protected attributes is duplicated, I moved it to the `shouldProtectAttribute` function, which is only executed when `isFirstPass` is truthy. It's worth noting that I'm not passing `LView` directly, but instead have to pass `isFirstPass` in different places. When I tried passing `LView` directly to `_setAttributeImpl`, I faced many circular reference errors.

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1673,7 +1683,19 @@ export function elementAttributeInternal(
   }
   const element = getNativeByTNode(tNode, lView) as RElement;
-  setElementAttribute(lView[RENDERER], element, namespace, tNode.value, name, value, sanitizer);
+  const isFirstPass = (lView[FLAGS] & LViewFlags.FirstLViewPass) === LViewFlags.FirstLViewPass;
```
I'm wondering if we can instead just pass `lView` into the `setElementAttribute` function (and similarly, into the function that sets properties) and retrieve the necessary info inside of the function (we'll have less logic for non-hydration cases). In this case we can avoid the following fields:

- isFirstPass flag
- hydration info

Both of those fields are only needed for hydration code paths, so passing `lView` would result in less work for non-hydration cases.

We can also consider passing `tNode` (or `null` where it's not available) instead of an index -- that can help simplify the contract of the set property and set attribute functions.

AndrewKushnir replied:
> When I tried passing LView or lView[FLAGS] (requires to import LViewFlags) directly to _setAttributeImpl, I faced many circular reference errors.

Could you please share an example? We can potentially refactor some code to avoid circular references (I guess we can import some symbols as types, e.g. `import {type LView, ...`).

arturovt commented:
@AndrewKushnir made changes according to your comment.

rhutchison commented:
> @AndrewKushnir made changes according to your comment.

When this ships, I will celebrate for all of you

rhutchison commented:
@AndrewKushnir what's left?

AndrewKushnir commented:
@arturovt @rhutchison Sorry for the delay. I've added this fix to the v19 candidate list and we'll perform code review and testing as a part of v19 preparation process.

AndrewKushnir commented:
@arturovt thanks for working on this fix 👍 

We've were reviewing this change today and the conclusion was that this fix would work (and I remember that I was proposing this/similar approach earlier), but it introduces quite a lot of complexity and runtime code at perf-sensitive locations, which would be very hard to manage. @alxhub proposed a different and a **very interesting** approach. Instead of relying on the overwritten behavior at runtime, we can serialize more info on the server and pre-populate corresponding LView slots with the necessary attribute values during hydration. In this case, a non-hydration runtime code would not require any changes - an update of an attribute would not happen with an existing runtime logic.

Here are some extra details that might be helpful for this change.

1. The serialization logic can be added to the `serializeLView` function [here](https://github.com/angular/angular/blob/main/packages/core/src/hydration/annotate.ts#L615), you can take a look how TNodes with the `TNodeType.Element` type are processed. The TNode would contain a tag name (`tNode.value`) and you'd also need an attribute/property name and value - this would require some extra code, see points below. 

    - Static attributes (e.g. <iframe src="...">): those attributes (both name and value ) would be stored on a TNode at the `tNode.mergedAttrs` slot.
    - Bindings (e.g. <iframe [src]="...">): the _value_ of an attribute/property would be stored in an `LView`, but the _name_ is currently not stored in **production** mode, only in dev mode by using the `storePropertyBindingMetadata` function. We may potentially need a similar function (or reuse the same one?) and invoke it only when we are on the server. We can potentially bring this logic in in a similar way how it's done for attribute and property setter functions in this PR.

2. The necessary information (a list of key-value pairs) can be included into the serialized output (it'd be present in the `nghData` in the TransferState). For example:

```
nghData: {
  ...
  {  // this object represents an LView info:
    attributes: {
      2 /* element idx */: [5 /* binding idx */, 'src', '...'],
      7 /* element idx */: [-1 /* `-1` means static */, 'src', '...'],
      ...
    } 
  }
}
```

The idea is that we'll have a special object in a serialized object for each `LView` and we can use an element LView slot id as an index and represent key-value pairs as a flat array + include binding slot index (we can also optimize it and have a string with separators as a value, e.g. `2|src|https://some-url.com`).

3. On the client, during hydration (in the `locateOrCreateElementNodeImpl` function), we can do the following:
    - check to see if we an element with a given index have special attributes
    - if yes, read the data and determine whether it's a static value or a dynamic binding
    - for static values - we'll use it to skip a "write" operation in the `setupStaticAttributes` (see point 4 below)
    - for dynamic bindings - restore it in `LView[index]`

4. For static values, we may need to do an extra update to the `setupStaticAttributes` function to avoid resetting the value (for attributes that were serialized as "special"). There are a few ways to do it, we can look into that if the dynamic binding case works correctly.

Please let me know if you need any extra info or have any questions, we'll be happy to help. Also, if you'd like, I can create an initial draft of the change and share with you (in this thread) as a diff.

Let's also keep this PR open for now, we'll get back to it in case the described approach doesn't work.

Thank you.

arturovt commented:
@AndrewKushnir okay, I'll need to go through the functions you mentioned locally in the playground to see what data they're working with and how they're serializing the LView.
